druntime/src/core/internal/array/concatenation.d

```diff
@@ -8,71 +8,160 @@
 */
 module core.internal.array.concatenation;
 
-/// See $(REF _d_arraycatnTX, rt,lifetime)
-private extern (C) void[] _d_arraycatnTX(const TypeInfo ti, scope byte[][] arrs) pure nothrow;
+private template CatElems(string catSingleElem, string catArray)
+{
+    const char[] CatElems = q{
+        foreach (ref from; froms)
+            static if (is (typeof(from) : T))
+                } ~ catSingleElem ~ q{
+            else
+            } ~ "{\n" ~ catArray ~ "\n}\n";
+}
 
-/// Implementation of `_d_arraycatnTX` and `_d_arraycatnTXTrace`
-template _d_arraycatnTXImpl(Tarr : ResultArrT[], ResultArrT : T[], T)
+Tret _d_arraycatnTX(Tret, Tarr...)(scope auto ref Tarr froms) @trusted
 {
-    private enum errorMessage = "Cannot concatenate arrays if compiling without support for runtime type information!";
-
-    /**
-    * Concatenating the arrays inside of `arrs`.
-    * `_d_arraycatnTX([a, b, c])` means `a ~ b ~ c`.
-    * Params:
-    *  arrs = Array containing arrays that will be concatenated.
-    * Returns:
-    *  A newly allocated array that contains all the elements from all the arrays in `arrs`.
-    * Bugs:
-    *  This function template was ported from a much older runtime hook that bypassed safety,
-    *  purity, and throwabilty checks. To prevent breaking existing code, this function template
-    *  is temporarily declared `@trusted pure nothrow` until the implementation can be brought up to modern D expectations.
-    */
-    ResultArrT _d_arraycatnTX(scope const Tarr arrs) @trusted pure nothrow
+    import core.internal.traits : hasElaborateCopyConstructor, Unqual;
+    import core.lifetime : copyEmplace;
+    import core.stdc.string : memcpy;
```

Comment:
Note that using memcpy here makes it impossible to use this hook in betterC mode.

Reply:
It's not necessarily a blocker for this PR, but it will require a future pass to break the dependency on c's std library.

Reply:
But is array concatenation allowed with `-betterC`, since dynamic arrays are disabled?

Reply:
it depends if you use the gc or not. if we are able to break the dependencies on the rest of druntime it would be a great win to use array cat in betterC code.

Reply:
Sounds good, but how could I not use the GC? Setting `res.length = totalLen;` below calls `_d_arraysetlengthT` ([1] that calls `rt.lifetime._d_arraysetlengthT()` [2], which allocates memory using `__arrayAlloc()` [3]), which involves the GC. And conceptually, array concatenation implies allocating memory for the resulting array, which can either be done via the GC or using libc via `malloc()`. Is there any other way?

[1] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/core/internal/array/capacity.d#L37-L53
[2] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/rt/lifetime.d#L1488
[3] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/rt/lifetime.d#L427-L447

Reply:
`memcpy` is part of the C runtime library, not druntime, so it is available in betterC